torch/csrc/distributed/c10d/c10d_frontend.cpp

```diff
@@ -0,0 +1,635 @@
+}
+
+void DistributedC10d::send(
+    at::Tensor tensor,
+    int64_t dst,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  auto work = isend(std::move(tensor), dst, group, tag);
+  if (work) {
+    work->wait();
+  }
+}
+
+int64_t DistributedC10d::recv(
+    at::Tensor tensor,
+    const c10::optional<int64_t>& src,
+    const std::shared_ptr<ProcessGroup>& group,
+    c10::optional<int64_t>& tag) {
+  if (rankNotInGroup(group)) {
+    return -1;
+  }
+
+  std::vector<at::Tensor> outputs = {std::move(tensor)};
+  if (!src.has_value()) {
+    auto work = group->recvAnysource(outputs, tag.value_or(0));
+    work->wait();
+    auto src_rank = work->sourceRank();
+    if (group == default_pg_) {
+      return src_rank;
+    }
+
+    return getGlobalRank(group, src_rank);
+  }
+
+  if (group == default_pg_) {
+    group->recv(outputs, src.value(), tag.value_or(0))->wait();
+  } else {
+    int64_t group_src_rank = getGroupRank(group, src.value());
+    group->recv(outputs, group_src_rank, tag.value_or(0))->wait();
+  }
+
+  return src.value();
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::broadcastMultiGPU(
+    std::vector<at::Tensor>& tensor_list,
+    int64_t src,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op,
+    int64_t src_tensor) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  BroadcastOptions opts;
+  opts.rootRank = src;
+  opts.rootTensor = src_tensor;
+
+  checkDefaultPg();
+  std::shared_ptr<ProcessGroup::Work> work;
+  if (group == default_pg_) {
+    work = default_pg_->broadcast(tensor_list, opts);
+  } else {
+    int64_t group_src_rank = getGroupRank(group, src);
+    opts.rootRank = group_src_rank;
+    work = group->broadcast(tensor_list, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::broadcast(
+    at::Tensor tensor,
+    int64_t src,
+    const std::shared_ptr<ProcessGroup>& group,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  BroadcastOptions opts;
+  opts.rootRank = src;
+  opts.rootTensor = 0;
+
+  std::vector<at::Tensor> tensors = {std::move(tensor)};
+  std::shared_ptr<ProcessGroup::Work> work;
+  checkDefaultPg();
+  if (group == default_pg_) {
+    work = group->broadcast(tensors, opts);
+  } else {
+    int64_t group_src_rank = getGroupRank(group, src);
+    opts.rootRank = group_src_rank;
+    work = group->broadcast(tensors, opts);
+  }
+
+  if (async_op) {
+    return work;
+  }
+  work->wait();
+  return nullptr;
+}
+
+std::shared_ptr<ProcessGroup::Work> DistributedC10d::allReduceMultiGPU(
+    std::vector<at::Tensor>& tensor_list,
+    const std::shared_ptr<ProcessGroup>& group,
+    ReduceOp op,
+    bool async_op) {
+  if (rankNotInGroup(group)) {
+    return nullptr;
+  }
+
+  AllreduceOptions opts;
+  opts.reduceOp = op;
+
```

Comment:
Why this does not check `group == default_pg_`, as broadcast did?

Reply:
and ditto for the following two allReduce* functions

Reply:
Because in here and some of following allReduce* functions, whether `group` is `default_pg_` or not doesn't affect the action taken. In Python we need this check because GroupMember.WORLD is not a group, but in this implementation, GroupMember.WORLD is simply a pointer to default_pg_.

In other words, if I do check for `group == default_pg_`, the logic would become

```
if (group == default_pg_) {
    default_pg->allreduce(tensor_list, opts)
} else {
    group->allreduce(tensor_list, opts)
}
```

both branches are logically identical.